src/Fl_Group.cxx

```diff
@@ -456,12 +511,25 @@ void Fl_Group::insert(Fl_Widget &o, int index) {
   if (o.parent()) {
     Fl_Group* g = o.parent();
     int n = g->find(o);
-    if (g == this) {
+    if (g == this) { // avoid expensive remove() and add() if we just move a widget within the group
       if (index > n) index--;
-      if (index == n) return;
+      index = on_move(n, index);
+      if (index == n) return; // this includes (children_ == 1)
+      if (index >= children_ || index < 0) return;
+      if (index > n)
+        memmove(array_+n, array_+(n+1), (index-n) * sizeof(Fl_Widget*));
+      else
+        memmove(array_+(index+1), array_+index, (n-index) * sizeof(Fl_Widget*));
+      array_[index] = &o;
+      init_sizes();
+      return;
     }
     g->remove(n);
   }
+
+  index = on_insert(&o, index);
+  if (index==-1) return;
```

Comment:
suggest: `if (index == -1) return;` for consistency and better readability (spaces around comparison operator).